src/io/hexdump.cr

```diff
@@ -0,0 +1,59 @@
+# IO object that prints an hexadecimal dump of all transfered data.
+#
+# Especially useful for debugging binary protocols on an IO, to understand
+# better when and how data is sent or received.
+#
+# By default `IO::Hexdump` won't print anything; you must specify which of
+# `read`, `write` or both you want to print.
+#
+# Example:
+# ```
+# require "io/hexdump"
+# io = IO::Hexdump.new(socket, output: STDERR, read: true)
+# ```
+#
+# When data is read from `io` it will print something akin to the following on
+# STDERR:
+# ```
+# 00000000  50 52 49 20 2a 20 48 54  54 50 2f 32 2e 30 0d 0a  PRI * HTTP/2.0..
+# 00000010  0d 0a 53 4d 0d 0a 0d 0a                           ..SM....
+# 00000000  00 00 00 04                                       ....
+# 00000000  00                                                .
+# 00000000  00 00 00 00                                       ....
+# ```
+module IO
+  class Hexdump
+    include IO
+
+    def initialize(@io : IO, @output : IO = STDERR, @read = false, @write = false)
+    end
+
+    def read(buf : Slice(UInt8))
+      @io.read(buf).tap do
+        @output.puts buf.hexdump if @read
```

Comment:
I didn't try this, and you've been using this a lot, but here the whole slice is printed. Shouldn't only the amount that was read be printed? Something like:

```cr
@io.read(buf).tap do |read_bytes|
  @output.puts buf.hexdump[0, read_bytes] if @read
end
```

?

The same question for `write`